CLN: Moving base client to the base backend

datapythonista opened the pull request:
Closes #2794

We have `ibis/backends/base` for the parent classes of what backends implement. But the base class for the `SQLClient` is in `ibis/client.py` instead, which is misleading and inconsistent.

Also, in that file there are unused things, and things unrelated to the client. In this PR I delete that file, remove the unused code, and moved the used one to a more appropriate location.

jreback commented on ibis/backends/tests/base.py:
```diff
@@ -123,7 +123,7 @@ def least(
 
     @property
-    def db(self) -> ibis.client.Database:
+    def db(self):
```
why are you removing typing?

datapythonista replied:
I updated the location of `Database`, and having annotations in tests doesn't seem so useful, so just deleted them.

I can restore them to the right location if you think they are important. But we don't have so many annotations in the code, and having those tests annotated is annoying when refactoring, so to meet they cause more harm than good. But not a big deal.

This PR is not very big, it's just that lots of imports had to be updated. Btw, #2809 is ready, and it's more useful to me to have it merge sooner, since it's blocking other refactorings of the compiler).

jreback reviewed (changes requested):
this is changing a lot
will have a closer look later

datapythonista commented:
This should be ready now. The main change here is moving `Client` and `SQLClient` to the appropriate locations in the base backend. And updating all the imports that use them.

There are few other changes (happy to discuss them if you think they are not a good idea):
- Removed type annotations of `Client` in tests, as these had to be updated, and didn't add much value IMHO. I think the only reason we had them is that they were added in a test conf, and all the rest are just from copy pasting the first (as too much code in Ibis, exactly why I'm trying to fix with the refactoring)
- Moved a function in `ibis/client.py` to find backends to a method of expression. This is the magic that makes `expr.execute()` call the execute in the backend.
- Removed `DatabaseEntity` which didn't look like it was being used, or useful
- Removed `DatabaseNamespace` which looked like a quite complex shortcut only present in Impala. I think what it achieves can be implemented in a much simpler way in the user code (it should be as simple as `prefix = 'whatever' ; con.table(prefix + 'table_name')`.

jreback commented:
> Moved a function in ibis/client.py to find backends to a method of expression. This is the magic that makes expr.execute() call the execute in the backend.

i think this is rather big change; e.g. we are all registered functions with dispatch now, this changes it to be more class based which is not the pattern.

datapythonista commented:
> i think this is rather big change; e.g. we are all registered functions with dispatch now, this changes it to be more class based which is not the pattern.

I think I didn't explain myself well. This should be an arbitrary change. The class `Expr` has methods `compile` and `execute`, that called methods in `ibis/client.py` that imported a `find_backends`function. That didn't feel very natural, it seemed to make more sense if the `find_backends` was a private method of `Expr`, and `execute` and `compile` could use it from there.

This shouldn't affect the dispatch system, which I didn't touch, and tests continue to pass, so seems like it's working exactly as before.

Does this make sense? Or am I missing something?

datapythonista commented:
@jreback all green after the changes and rebase

jreback commented on ibis/expr/types.py:
```diff
@@ -188,6 +190,55 @@ def op(self):
     def _factory(self):
         return type(self)
 
+    def _find_backends(self):
```
next version can you type the outputs here

datapythonista replied:
> next PR can you type the outputs here

I think it's worth making mypy run properly before spending much time in annotating new things. I created #2823, I'll see if I can spend some time on it once I finish the bigger part of the refactoring, which is almost over.

jreback commented:
thanks @datapythonista
